compiler/rustc_middle/src/middle/effective_visibility.rs

```diff
@@ -0,0 +1,133 @@
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def_id::LocalDefId;
+use rustc_middle::middle::privacy::AccessLevel;
+use rustc_data_structures::stable_hasher::{HashStable, StableHasher};
+use rustc_macros::HashStable;
+use rustc_query_system::ich::StableHashingContext;
+use crate::ty::Visibility;
+use crate::ty::DefIdTree;
+
+#[derive(Debug, Clone)]
+pub struct EffectiveVisibilities {
+    pub map: FxHashMap<LocalDefId, EffectiveVisibilityValue>,
+}
+
+#[derive(Clone, Debug, Copy, HashStable)]
+pub struct EffectiveVisibilityValue {
+    vis: [Option<LocalDefId>; 4]
+}
+
+impl IntoIterator for EffectiveVisibilityValue {
+    type Item = Option<LocalDefId>;
+    type IntoIter = std::array::IntoIter<Self::Item, 4>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.vis.into_iter()
+    }
+}
+
+impl EffectiveVisibilityValue {
+    pub fn new() -> EffectiveVisibilityValue {
+        EffectiveVisibilityValue { vis: Default::default() }
+    }
+
+    fn get_idx(&self, tag: AccessLevel) -> usize {
+        match tag {
+            AccessLevel::Public => 0,
+            AccessLevel::Exported => 1,
+            AccessLevel::Reachable => 2,
+            AccessLevel::ReachableFromImplTrait => 3
+        }
+    }
+
+    pub fn get(&self, tag: AccessLevel) -> Option<&LocalDefId> {
+        let idx = self.get_idx(tag);
+        self.vis[idx].as_ref()
+    }
+
+    pub fn get_mut(&mut self, tag: AccessLevel) -> Option<&mut LocalDefId> {
+        let idx = self.get_idx(tag);
+        self.vis[idx].as_mut()
+    } 
+
+    pub fn set(&mut self, vis_id: LocalDefId, tag: AccessLevel) {
+        let idx = self.get_idx(tag);
+        self.vis[idx] = Some(vis_id);
+    }
+
+    pub fn min(&self, tag: AccessLevel) -> Option<LocalDefId> {
+        let mut idx: isize = 4;
+        let id = self.into_iter().rev().find(|&x|{
+            idx -= 1;
+            x.is_some() && (idx <= self.get_idx(tag) as isize )
+        });
+        match id {
+            Some(res) => res,
+            None => None
+        }
+    }
+
+    pub fn merge_with<T: DefIdTree>(&mut self, def_id: LocalDefId, tag: AccessLevel, tree: T) {
+        if let Some(current_value) = self.get(tag) {
+            if tree.is_descendant_of(def_id.to_def_id(), current_value.to_def_id()) {
+                self.set(def_id, tag);
+            }
+        } else {
+            self.set(def_id, tag);
+        }
+    }
+
+    pub fn calculate_effective_vis<T: DefIdTree>(
```

Comment:
The less code lives in `rustc_middle` the better.
This logic is only used in `compiler/rustc_resolve/src/access_levels.rs` so it can be moved there.